lib/shelley/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs

```diff
@@ -2795,27 +2771,47 @@ instance Buildable a => Show (ShowBuildable a) where
 instance Arbitrary (Hash "Datum") where
     arbitrary = pure $ Hash $ BS.pack $ replicate 28 0
 
+instance Arbitrary (Cardano.AddressInEra Cardano.AlonzoEra) where
+    arbitrary = genAddressInEra Cardano.AlonzoEra
+
+instance Arbitrary (Cardano.TxOutValue Cardano.AlonzoEra) where
+    arbitrary = Cardano.TxOutValue Cardano.MultiAssetInAlonzoEra
+        <$> genValueForTxOut
+    shrink _ = [] -- TODO
+
+instance Arbitrary (Cardano.TxOutDatum ctx Cardano.AlonzoEra) where
+    arbitrary = genTxOutDatum Cardano.AlonzoEra
+
+instance Arbitrary (Cardano.TxOut ctx Cardano.AlonzoEra) where
+    arbitrary = genTxOut Cardano.AlonzoEra
+    shrink (Cardano.TxOut addr val dat) = tail
+        [ Cardano.TxOut addr' val' dat'
+        | addr' <- prependOriginal shrink addr
+        , val' <- prependOriginal shrink val
+        , dat' <- prependOriginal shrink dat
+        ]
+
 instance Arbitrary (PartialTx Cardano.AlonzoEra) where
     arbitrary = do
         let era = AlonzoEra
         tx <- genTxForBalancing era
         let (Cardano.Tx (Cardano.TxBody content) _) = tx
         let inputs = Cardano.txIns content
-        resolvedInputs <- forM inputs $ \i -> do
+        resolvedInputs <- Cardano.UTxO . Map.fromList <$> forM inputs (\i -> do
```

Comment:
Suggestion: Use `BlockArguments`, as it would allow us to remove the parentheses from around the monadic expression:

```suggestion
        resolvedInputs <- Cardano.UTxO . Map.fromList <$> forM inputs \i -> do
```

We would then not need to "smuggle in" the trailing bracket a few lines later.

Reply:
Unrelated: @jonathanknowles , could you record this example as an argument in favor of `BlockArguments`? Personally, I'm not yet convinced by this extension (does `Cardano.UTxO . Map.fromList <$> forM inputs \i -> i` parse?) — regardless, I think that the use or unuse of `BlockArguments` should be discussed with the team so that we have a common Haskell idiom that we share.

(One reason for me insisting on a shared Haskell idiom is past use of lenses in the wallet — if I understood correctly, that didn't work out so well, and I believe that's because they were not explicitly made part of a shared idiom. (Even more unrelated — happy to outside of this PR).)

Reply:
🤔 Ah, thanks! But I will not apply this before there is team consensus then.

> Does `Cardano.UTxO . Map.fromList <$> forM inputs \i -> i parse?`

No:

<img width="831" alt="Skärmavbild 2022-05-13 kl  16 21 12" src="https://user-images.githubusercontent.com/304423/168304142-68f527c1-2348-4006-a9d2-b27b37dde7fd.png">

Often you can avoid superfluous parenthesis by flipping the order of things, but I didn't find that to be possible here (and a few similar cases I have run into)

Reply:
> (does Cardano.UTxO . Map.fromList <$> forM inputs \i -> i parse?

@HeinrichApfelmus here is a commit that compiles: https://github.com/input-output-hk/cardano-wallet/commit/bcd3b4541112020f98700af332247f470ca01b10

> I think that the use or unuse of BlockArguments should be discussed with the team so that we have a common Haskell idiom that we share.

Fair enough! I started a poll in our development channel [here](https://element.adrestia.iohkdev.io/#/room/!pVmOFJiBkItJMNTmCR:adrestia.iohkdev.io/$SUQD44U_30OQl_hpz2dZUGyBoRnz_8M9Ng6-FCN5cTQ?via=adrestia.iohkdev.io).

I personally don't have a strong feeling either way about `BlockArguments`, but I do think that `do` blocks can become a bit unwieldy (and difficult to follow) when they are are enclosed within parentheses spanning multiple lines. In this case, the trailing brace is rather difficult to spot in the GitHub viewer. (Though with a real editor, we'd have matching brace highlighting to help us.)

If we couldn't use `BlockArguments`, then I'd advocate for extracting out the "block" to make it a separate value in cases like these.

Reply:
> but I do think that do blocks can become a bit unwieldy (and difficult to follow) when they are are enclosed within parentheses spanning multiple lines.

The traditional solution to that is to have the `do` block be the last argument of the function, and slap a `$` before it, e.g.

```
foo arg1 arg2 $ \i -> do
    something
```

However, this solution does not work in this case, because of the fixity of the operator `<$>` is higher than the fixity of `$`, so `<$>` binds tigher. Put differently, the expression

```
Map.fromList <$> forM input $ \i -> do {…}
```

is parsed as

```
(Map.fromList <$> forM input) $ \i -> do {…}
```

but we want this:

```
Map.fromList <$> (forM input $ \i -> do {…})
```

My personal go-to solution is to use `fmap` instead of the operator:

```
fmap Map.fromList . forM input $ \i -> do {…}

```

> Fair enough! I started a poll in our development channel here

Thanks! My proposal would be to discuss and move this forward this during our next dev meeting as well. (It was on the agenda already, sorry for postponing.)

Reply:
> My personal go-to solution is to use fmap instead of the operator:

🤔 I thought I tried that! 👍